bindings/java/c/host.c

```diff
@@ -0,0 +1,580 @@
+
+        // get java method
+        method =
+            (*jenv)->GetStaticMethodID(jenv, host_class, java_method_name, java_method_signature);
+        assert(method != NULL);
+
+        // set java method params
+        assert(context != NULL);
+        jcontext_index = context->index;
+
+        jmsg = (*jenv)->NewDirectByteBuffer(jenv, (void*)msg, sizeof(struct evmc_message));
+        assert(jmsg != NULL);
+
+        // call java method
+        jobject jresult =
+            (*jenv)->CallStaticObjectMethod(jenv, host_class, method, jcontext_index, jmsg);
+        assert(jresult != NULL);
+
+        struct evmc_result* result_ptr =
+            (struct evmc_result*)(*jenv)->GetDirectBufferAddress(jenv, jresult);
+        assert(result_ptr != NULL);
+        result = *result_ptr;
+    }
+    return result;
+}
+
+static struct evmc_tx_context get_tx_context_fn(struct evmc_host_context* context)
+{
+    struct evmc_tx_context result;
+    const char java_method_name[] = "get_tx_context";
+    const char java_method_signature[] = "(I)Ljava/nio/ByteBuffer;";
+    JNIEnv* jenv = attach();
+    if (jenv != NULL)
+    {
+        jclass host_class;
+        jmethodID method;
+        jint jcontext_index;
+        // get java class
+        host_class = (*jenv)->FindClass(jenv, "org/ethereum/evmc/Host");
+        assert(host_class != NULL);
+
+        // get java method
+        method =
+            (*jenv)->GetStaticMethodID(jenv, host_class, java_method_name, java_method_signature);
+        assert(method != NULL);
+
+        // set java method params
+        assert(context != NULL);
+        jcontext_index = context->index;
+
+        // call java method
+        jobject jresult = (*jenv)->CallStaticObjectMethod(jenv, host_class, method, jcontext_index);
+        assert(jresult != NULL);
+
+        struct evmc_tx_context* result_ptr =
+            (struct evmc_tx_context*)(*jenv)->GetDirectBufferAddress(jenv, jresult);
+        assert(result_ptr != NULL);
+        result = *result_ptr;
+    }
+    return result;
+}
+
+
+static evmc_bytes32 get_block_hash_fn(struct evmc_host_context* context, int64_t number)
+{
+    evmc_bytes32 result;
+    char java_method_name[] = "get_code_hash";
+    char java_method_signature[] = "(IJ)Ljava/nio/ByteBuffer;";
+    JNIEnv* jenv = attach();
+    if (jenv != NULL)
+    {
+        jclass host_class;
+        jmethodID method;
+        jint jcontext_index;
+        jlong jnumber;
+
+        // get java class
+        host_class = (*jenv)->FindClass(jenv, "org/ethereum/evmc/Host");
+        assert(host_class != NULL);
+
+        // get java method
+        method =
+            (*jenv)->GetStaticMethodID(jenv, host_class, java_method_name, java_method_signature);
+        assert(method != NULL);
+
+        // set java method params
+
+        assert(context != NULL);
+        jcontext_index = context->index;
+
+        // FYI, a java long can hold a 64bit unsigned number in long by
```

Comment:
`int64_t` means it is signed

Reply:
lol...I knew that. no idea why i left that comment. I'll remove